### ISSUE TITLE
Add support for `pkg.exports.module` or similar

### ISSUE THREAD

**Issue by ctavan** [webpack/webpack#11014]
I would like to move the discussion that started in https://github.com/uuidjs/uuid/pull/462#issuecomment-639678696 and https://github.com/rollup/rollup/issues/3514#issuecomment-640014379 into this project.

**TL;DR:** We do believe that, apart from support for `pkg.exports.browser` that is being added in https://github.com/webpack/webpack/pull/10953, we may need to support an additional key in the `pkg.exports` object.

Let's illustrate this with an example taken from [`uuid`](https://github.com/uuidjs/uuid/blob/0f6c4365275e1f7ef9ee875e01415cd466ae977a/package.json#L20-L34). Before `pkg.exports` we used to specify the following in package.json:

```json
  "main": "./dist/index.js",
  "module": "./dist/esm-node/index.js",
  "browser": {
    "./dist/md5.js": "./dist/md5-browser.js",
    "./dist/rng.js": "./dist/rng-browser.js",
    "./dist/sha1.js": "./dist/sha1-browser.js",
    "./dist/esm-node/index.js": "./dist/esm-browser/index.js"
  },
```

The goal was the following:

* Node.js native should pick `main`
* Bundlers targeting Node.js should pick `module`
* Bundlers targeting the browser would pick `"module": "./dist/esm-node/index.js",` but then use the override from `browser`: `"./dist/esm-node/index.js": "./dist/esm-browser/index.js"`
* (Before we added the esm build, bundlers targeting the browser would have used the CommonJS build from `main`, hence the additional browser-overrides)

Now that Node.js has added and webpack is about to add `pkg.exports` support the above can no longer be achieved for webpack:

```json
  "main": "./dist/index.js",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "browser": "./dist/esm-browser/index.js",
      "require": "./dist/index.js",
      "import": "./wrapper.mjs"
    }
  },
  "module": "./dist/esm-node/index.js",
  "browser": {
    "./dist/md5.js": "./dist/md5-browser.js",
    "./dist/rng.js": "./dist/rng-browser.js",
    "./dist/sha1.js": "./dist/sha1-browser.js",
    "./dist/esm-node/index.js": "./dist/esm-browser/index.js"
  },
```

Since Node.js potentially suffers from the [dual package hazard](https://nodejs.org/docs/latest-v14.x/api/esm.html#esm_dual_package_hazard) `uuid` provides a `./wrapper.mjs` which wraps the CJS build and exports it as ESM for `pkg.exports.import`.

While the browser use case is covered by `pkg.exports.browser` for webpack, things no longer work as expected when bundling for Node.js

While it would technically work to just fall back to `pkg.exports.import` and `pkg.exports.require` that is at least not an option for ESM-only bundlers like rollup, see https://github.com/rollup/rollup/issues/3514.

Hence the need for one additional `pkg.exports` key that allows specifying a "pure esm" build to be used for Node.js bundling.

For more details see the discussion in https://github.com/uuidjs/uuid/pull/462#issuecomment-639678696 onwards.

/cc @sokra @lukastaegert @guybedford @jkrems @evilebottnawi

**Comment by guybedford** [comment]
I still disagree with the premise for such a feature, based on the following arguments:

1. If you have a bundler build a separate conditional path to the environment itself, this goes against a major design goal of bundlers in that they need to retain the semantics of the target environment as they perform optimization. To get a different behaviour of code when bundled versus unbundled should not be a goal.
2. Further, the dual mode hazard applies equally to the scenario where users are bundling for Node.js. Because when bundling multiple packages, any other dependency might still use a require version of uuid thereby exposing the hazard. Thus the same hazard argument applies to bundling as it does in Node.js, and selectively deciding that this hazard does not matter for bundling seems arbitrary.

I would suggest two possible paths forward:

1. Either decide that the hazard does not matter for your users and ship a pure esm build
2. Decide that it does matter and enforce these same CJS semantics in the bundling process

Any attempt to find a middle path here will lead to issues

* Creating patterns that separate bundler semantics which not only creates an issue in this case, but also creates a pattern that will create these issues for other users and encourage creating these issues for other users when bundling libraries and finding surprising semantics changes that users cannot control
* Complicating the already complex knowledge required to process how to publish packages with "exports" support when the major goal right now should be simplifying the communication discussion and shared ideas around this

For these reasons I remain strongly against further work here.

**Comment by ctavan** [comment]
> 1. If you have a bundler build a separate conditional path to the environment itself, this goes against a major design goal of bundlers in that they need to retain the semantics of the target environment as they perform optimization. To get a different behaviour of code when bundled versus unbundled should not be a goal.

Is that a major design goal? I have personally never seen it that way TBH.

> 2. Further, the dual mode hazard applies equally to the scenario where users are bundling for Node.js. Because when bundling multiple packages, any other dependency might still use a require version of uuid thereby exposing the hazard. Thus the same hazard argument applies to bundling as it does in Node.js, and selectively deciding that this hazard does not matter for bundling seems arbitrary.

Could you explain in more detail where the dual package hazard would exist in a bundler scenario?

With the current situation (`webpack@4` or `rollup` with `@rollup/plugin-node-resolve` and `@rollup/plugin-commonjs`) as a package maintainer I have a _guarantee_ that both bundlers will _always_ pick the ESM build specified by the `pkg.module` field, even if nested dependencies use `require()` instead of `import`. This effectively eliminates the hazard that exists in Node.js because I'm _guaranteed_ I will get precisely _one_ build of my package (Ignoring the case of incompatible semver ranges where multiple _versions_ of the same package have to be included. This problem equally applies to Node.js and has always existed!).

This practice has never produced any real world trouble for me in the past, I'd be curious to learn about cases where the current situation produces problems.

If we follow your suggestion and remove support for what `pkg.module` currently gives us (thus bringing Node.js bundles closer to how they would be treated in Node.js unbundled), then we'll immediately run into https://github.com/rollup/rollup/issues/3514, and this would question the very fundamental principle of `rollup` of being a _pure ESM_ bundler by default. I believe @lukastaegert has an opinion on that.

I'm not entirely rejecting your suggestion to treat Node.js sources when bundling them in the same way as Node.js would treat them. In practice that would just mean for me that I'd need rollup to support https://github.com/rollup/rollup/issues/3514 out-of-the box.

**Comment by sokra** [comment]
> 2. Further, the dual mode hazard applies equally to the scenario where users are bundling for Node.js. Because when bundling multiple packages, any other dependency might still use a require version of uuid thereby exposing the hazard. Thus the same hazard argument applies to bundling as it does in Node.js, and selectively deciding that this hazard does not matter for bundling seems arbitrary.

Dual mode hazard doesn't exist for any bundler I know of. This is something node.js created by resolving based whether you are using `require()` or `import` to load a package. In webpack resolving is independent of that and you always get the same version (may be either ESM or CJS, we don't mind).

When adding support for the `exports` field and `require` and `import` we would now inherit the possibility from the node.js semantics.

> 1. If you have a bundler build a separate conditional path to the environment itself, this goes against a major design goal of bundlers in that they need to retain the semantics of the target environment as they perform optimization. To get a different behaviour of code when bundled versus unbundled should not be a goal.

Yes, package author could provide a module with different semantics in the `exports.module` field. But why should they do so. And this is not a new problem, the same is true for the `module` field or the `browser` field.

**Comment by jkrems** [comment]
> Either decide that the hazard does not matter for your users and ship a pure esm build

FWIW: I strongly suspect this is effectively the case for UUID. The only scenario where it matters (afaict) is when UUID is used by two different importers, one using `require`, one using `import`, both using v1 UUIDs but not passing `node`/`clocksq`, and them relying on getting the same ephemeral UUID namespace for this particular thread. Which sounds like a _very_ narrow use case, if it actually exists. And even then it could potentially be solved in the future by using a JSON module to share state, removing the need for CJS in the ESM path.

All that being said: Having a special "bundle only" or "module only" condition that can be used in the two cases brought up so far (bundler allows `require(esm)`, bundle doesn't support `import(cjs)`) doesn't sound awful. It certainly seems like a niche feature but it may be worth it for some packages. I share @guybedford concerns though that it creates a scenario where bundlers actively run different code than would run unbundled. But that's already the case in a `module`-field world so maybe it's too late to change that.

**Comment by ctavan** [comment]
> > Either decide that the hazard does not matter for your users and ship a pure esm build
> 
> FWIW: I strongly suspect this is effectively the case for UUID. The only scenario where it matters (afaict) is when UUID is used by two different importers, one using `require`, one using `import`, both using v1 UUIDs but not passing `node`/`clocksq`, and them relying on getting the same ephemeral UUID namespace for this particular thread. Which sounds like a _very_ narrow use case, if it actually exists. And even then it could potentially be solved in the future by using a JSON module to share state, removing the need for CJS in the ESM path.

Just FYI, I'm [well aware](https://github.com/uuidjs/uuid/pull/402#discussion_r389406987) that the dual package hazard is rather hypothetical and probably not a problem in practice.

A JSON module to share state would be a very interesting option if supported by `rollup`, after all the fact that importing `.cjs` is not supported by `rollup` out of the box led me to [discard the state isolation approach](https://github.com/uuidjs/uuid/pull/402#issuecomment-620073703) and instead [use the wrapper approach for Node.js](https://github.com/uuidjs/uuid/pull/423).

**Comment by guybedford** [comment]
@ctavan all RollupJS users need to support CommonJS in bundling workflows. I think trying to work around this makes little sense.

If you consider the browser as a JS runtime, then the `"module"` field is very much supported for this runtime as a convention.

**Comment by guybedford** [comment]
But in terms of a path forward for uuid, I think @jkrems may well be correct in his assessment here that the dual mode hazard may not apply to uuid due to a lack of internal state that risks any conflict.

**Comment by ctavan** [comment]
> @ctavan all RollupJS users need to support CommonJS in bundling workflows. I think trying to work around this makes little sense.

Devil's in the detail: e.g. when using the state isolation approach with a `.cjs` file that indeed will work using `@rollup/plugin-commonjs` however you would currently have to [configure `.cjs` manually](https://github.com/rollup/rollup/issues/3514) as an extension to be matched by that plugin (or rollup would have to change the plugin config defaults). I firmly believe that shipping dependencies which require specific bundler options should really not be anyone's goal.

Rollup would have to adapt quite a bit to enable a seamless developer experience if more CommonJS interop is desired out-of-the-box. I'll leave it to @lukastaegert et. al. to decide if that's the desired direction for rollup.

> But in terms of a path forward for uuid, I think @jkrems may well be correct in his assessment here that the dual mode hazard may not apply to uuid due to a lack of internal state that risks any conflict.

As commented above I totally agree and I'm less concerned about `uuid` where we definitely have simple options at hand. I rather hope that using `uuid` as an example we might be able to clarify some other questions around `pkg.exports` that still seem to be unclear.

**Comment by guybedford** [comment]
I was not aware that the `.cjs` extension was not supported by default in the CommonJS plugin. I agree this is an oversight and something we should address. //cc @lukastaegart

Continuing to improve the conventions here between tools to make these developer experiences more seamless is certainly a goal I agree, and thanks for helping drive these discussions.

What other ways do you think the RollupJS CommonJS plugin needs to adapt here? Perhaps we can start to split these off into their own issues.

**Comment by ctavan** [comment]
> What other ways do you think the RollupJS CommonJS plugin needs to adapt here? Perhaps we can start to split these off into their own issues.

I think it's really all about the way in which rollup treats CommonJS and I have described my concerns regarding `uuid` in https://github.com/rollup/rollup/issues/3514 (which was rejected for now).

**Comment by lukastaegert** [comment]
My position on this is that the we need something as a replacement for `package.module` for the future, especially if we want to make using `package.exports` the default. Let me elaborate:

## The status quo: `package.module`

At the moment, many libraries support a special `.module` property in their `package.json` file that offers an ESM build of that library. This attribute is currently used by bundlers as well as new "no-bundler" dev tools like Vite, Snowpack or es-dev-server for different reasons:
- only with pure ESM builds, bundlers can perform advanced optimizations such as tree-shaking and scope-hoisting
- only pure ESM modules can be loaded directly by browsers, which is what no-bundler-dev-tools rely on

## Why separate builds via `exports.import` and `exports.require` can be a bad thing for bundlers

Now for many libraries, we can easily transition to just using `package.exports` with branches for `import` and `require`. The devil however here is in the details, and I am not yet talking about stateful libraries:

If we decide to strictly follow `package.exports` semantics, then if in a bundler module graph the same library is both imported by an ES module as well as required by a CJS module, then this library will end up twice in the module graph: The ESM build and the CJS build. That is what the semantics imply. And it does not end here: If those libraries themselves have dependencies that again have separate ESM and CJS builds, then again those libraries will end up twice in the module graph.

This blows up the generated bundle completely needlessly as of course the functionality of those separate ESM and CJS build is the same. And I say "of course" even though there is no of course: There is nothing forcing the user to make sure that the ESM and CJS files have anything in common. And that is why when following the semantics strictly, bundlers NEED to pull in libraries twice. Otherwise they are generating meta-rules as to how CJS and ES builds of a library need to be compatible to be able to bundle that library that do not exist in the original semantics.

And of course if the library itself has state, it will likely be broken when pulled in twice. This is also known as the "dual package hazard".

## Why using "isolated state" or an "ESM wrapper" are a bad thing

Now one solution to solve the above issues is to make the ESM and CJS builds "share code". The "isolated state" approach will just share the state, so it will not really solve the problem of blown up bundles. The "ESM wrapper" approach solves both "in a way" as the majority of the code is shared by the two builds.

Both solution have in common that the ESM part at some point imports a CJS module with the shared code. And it cannot be the other way around because only ES modules can synchronously import CJS code—the opposite direction is only possible via asynchronous dynamic `import()` which makes the whole library asynchronous, which users will probably not like, especially if it only does some simple calculations.

This means that such a library is unusable by no-bundler-dev-tools, they will need to transpile it first to ESM. Which is a messy and imprecise process, often facilitated by `@rollup/plugin-commonjs`.

But also bundlers will produce sub-optimal results. So a library such as

```js
export const foo = 'Hello complicated world!';
```

can easily become something akin to

```js
var fooNamespace = createCommonjsModule(function (module) {
{
    exports.foo = () => 'Hello complicated world!';
}
});
```

So why is that bad? The thing is, it is much harder to identify named exports and do tree-shaking through such a wrapper function call. In the end even if some exports of a library are unused, with the wrapper approach, all functions will very likely end up in the bundle. And there is also scope-hoisting that is prevented by this wrapper. Scope-hoisting is a process where code from different modules is lifted into a shared module scope. This has two advantages: The result is slightly smaller, but it is also faster because
- runtime execution does not need to step in and out of functions when loading modules
- variables do not need to be picked off a "module.exports" object, avoiding the cost of unneeded property accesses

## How was this avoided by `package.module`

With `package.module`, there was always a "hidden contract", namely that the ESM build provided here is "sufficiently compatible" with the CJS build. I will talk about that contract later. It meant that for such packages, bundlers would resolve both ESM imports and CJS requires of that library against the ESM build, avoiding any duplication and allowing for further optimizations as outlined above.

## Conclusion

So in my opinion, moving to `exports` in its current form is likely a foot gun and a big step backwards for the two use-cases I am discussing here.

As a side-note, I also think that the suggestion of shared CJS state and ESM wrappers makes sure that in the Node world, ESM will never really catch on because, well, it does not solve a problem. But it will shackle many libraries to CJS. If it was possible to provide CJS wrappers for ESM instead where the ESM build is pure, well, then this would paint a bright future where the majority of Node libraries as well could eventually transition to ESM.

I am definitely open to suggestions for how to solve this, especially @guybedford seems to be very much against my conclusions, and I hope he has considered the points I listed above. I am saying if there will be no solution that solves these issues as least as good, then `package.module` will never die.

## My suggestion

This would be just one way of tackling it, and it is not really my suggestion but @sokra's, I am more thinking about the details. The suggestion is to add a now conditional export field, I will call it `moduleOnly`. E.g. a library could have

```
{
  "exports": {
    "moduleOnly": {
      // paths for consumers that prefer "modules only"
    },
    "import": ...
    "default": ...
  }
}
```

`moduleOnly` will have the following implications:
- All modules imported here are ES modules, and if they import a different library, it has either only an ESM build, an ESM build available via the "import" conditional export that is "pure ESM", or again a "moduleOnly" conditional export
- It provides the same paths and exports as a regular "import" conditional export if it is available
- If it has a CJS conditional export, then it provides again the same paths and all packages are "sufficiently compatible" (again I will discuss this below).

This has the following advantages of the previous `package.module`:
- It allows for "deep imports"
- It allows for browser branching or development branching via nested conditional exports
- These expectations are clearly defined

Note that I do not imply every package author has to know about this or use this. But they can add this as an optimization for bundlers and no-bundler-dev-tools

## What does it mean for a ESM and CJS to be compatible

This is likely stuff for some discussion. If a library exports just an object as CJS, then having is properties as named exports in ESM appears natural and requiring it should return the namespace.

But what to do about small libraries that only export e.g. one function? In CJS; this function would be assigned to `module.exports`. However, the most natural translation to ESM would be a package with a single default export. Incidentally, this is what Rollup does in inverse when converting from ESM to CJS in "auto" exports mode:
- If there are only named exports, the CJS exports are an object with the corresponding properties (named exports mode)
- If there is only a single default export, CJS assigns it to `module.exports` (default export mode)
- If there are mixed named and default exports, a warning is displayed and we switch to named export mode.

So my suggestion is to do that in reverse: requiring an ESM module should return
- Its default export if it exists and there are no other exports
- The namespace otherwise

I think (but please correct me) Webpack currently always returns the namespace. I find this problematic however as it would force authors of small libraries into unnatural interfaces such as `require('lib').default` for ALL their users if they want CJS and ESM to be compatible.

So in all, these are my thoughts, other suggestions and feedback welcome. Please note though that as of today I am on vacation and a family trip so I will only have little time to respond and likely not as detailed as above.

**Comment by guybedford** [comment]
When bundling for the browser, the `"exports"` `"browser"` field can allow always pointing to an ES module. This is actually superior to the current module field, because it in addition can allow branches such as `"development + browser"` and `"production + browser"`. It is also recommended for users to apply conditions such as `"react-native"` and `"deno"` even if they would like. All of these, so far as their target environment supports ESM, can support ESM.

But if you are bundling for Node.js, then certainly we are tied to CJS for 5 - 10 years at the very least. Denying that truth will cause a lot of pain.

**Comment by lukastaegert** [comment]
The problem is not a browser problem, it also exists to the same extent when bundling for Node. `acorn` for instance is a big library that I do no want to put in a bundle twice. Unfortunately it is both stateful and has plugins some of which are both CJS and import acorn. A `moduleOnly` condition would also solve this scenario that is very relevant for Rollup. Also people already pu all kinds of mixed stuff into the `browser` field and in the end the only contract is that stuff works in the browser.

**Comment by guybedford** [comment]
@lukastaegert there are just too many risks having code change when bundled versus unbundled. If I have a Node.js app that works one way and then bundle it there is such a high chance that some obscure ESM / CJS interop case will make things break. The only way to reliably bundle a Node.js application that is currently working correctly, is to have the bundling process for Node.js follow the exact same resolution semantics. Anything else is just .

This is specifically when bundling for Node.js only though, and if you want to build your library out of ES modules then publish a unified CJS build that does optimization + ESM -> CJS conversion that is very much the best practise.

Personally I would advise library authors to publish CommonJS from authored ESM, and then only publish an ESM version of their library if they are able to drop user support for Node.js < 12 and then go all-in on ESM rather.

The only other reason to do a dual package is if you have a CommonJS package and want to just provide a nice custom ESM facade for the package.

**Comment by lukastaegert** [comment]
> Personally I would advise library authors to publish CommonJS from authored ESM, and then only publish an ESM version of their library if they are able to drop user support for Node.js < 12 and then go all-in on ESM rather.

This will not happen if they want their library to remain relevant to Node users for the reasons I outlined, and well beyond Node 12 becomes unsupported.

We need a solution that is NOW as least as as good as package.module, otherwise I see no reason to make exports the highest priority when resolving packages.

**Comment by guybedford** [comment]
@lukastaegert I'm really not sure how to discuss this further without a clear argument to focus on. Perhaps we need a RollupJS meeting?

**Comment by lukastaegert** [comment]
I am on vacation and actually want to focus on that, but I was dragged into this. My goals and arguments are listed above.

**Comment by NMinhNguyen** [comment]
> I think (but please correct me) Webpack currently always returns the namespace.

I've actually run into a pretty annoying CJS/ESM interop issue (https://github.com/facebook/react/pull/18187 https://github.com/facebook/react/pull/18186#issuecomment-592794066) where webpack seemed to favour `package.module` even if it was being `require`d from a CJS module.

**Comment by sokra** [comment]
> So my suggestion is to do that in reverse: requiring an ESM module should return
> 
> * Its default export if it exists and there are no other exports
> * The namespace otherwise

> > I think (but please correct me) Webpack currently always returns the namespace.
> 
> I've actually run into a pretty annoying CJS/ESM interop issue ([facebook/react#18187](https://github.com/facebook/react/pull/18187) [facebook/react#18186 (comment)](https://github.com/facebook/react/pull/18186#issuecomment-592794066)) with a default export where webpack seemed to favour `package.module` even if my package was being `require`d from a CJS module. The package contained both ESM and CJS (via `package.main`) Rollup bundles. As described above, Rollup would generate `module.exports` without `.default` (the default named export mode), however webpack still seemed to prefer ESM.

I also noticed that rollup is behaving differently compared to webpack, babel, typescript and even to the [node PR](https://github.com/nodejs/node/pull/30891). I wonder why it doesn't align to existing tools.

It behaves different for require(ESM), but also when generating a CommonJS library from ESM.

A module `export default 42;` will result in:

``` js
// typescript
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = 42;
```

``` js
// babel
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _default = 42;
exports.default = _default;
```

``` js
// webpack (it's a bit more verbose, so I optimized it for readablility)
module.exports =
/******/ (() => { // webpackBootstrap
/******/ 	"use strict";
const exports = {};
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
Object.defineProperty(exports, '__esModule', { value: true });
Object.defineProperty(exports, key, { enumerable: true, get: () => __WEBPACK_DEFAULT_EXPORT__ });
const __WEBPACK_DEFAULT_EXPORT__ = (42);
/******/ 	return exports;
/******/ })();
```

``` js
// rollup
'use strict';

var main = 42;

module.exports = main;
```

On the importing side with `console.log(require("esm"));`. webpack, babel, typescript and the [node PR](https://github.com/nodejs/node/blob/0a1faa47ba6f0877ab7f9112fb272ea3118706e6/lib/internal/modules/cjs/loader.js#L1257) because similar to `import * as ns from "esm"; console.log(ns);` while with rollup it's getting pretty weird. The behavior depends on whether the esm has a default export or not. If it has one `require("esm")` returns the default export. Only the default export. Even if there are other named exports, there is no way to access these named exports from CommonJS. But if the esm doesn't contain a default export it actually behaves like the other tools and returns the namespace export. I also tried `require("esm").named` and `const { named } = require("esm")`, but these neither work. Even if that would work that would be pretty inconsistent.

So @NMinhNguyen you issue is caused by this inconsistency when generating a cjs library with rollup. You would not see this problem with `module` and `main` being different when `require()`ed when using babel, typescript, or webpack to transpile/bundle your library. And you could also workaround it by adding an additional named export next to the `export default` in the entrypoint. Yes this would require `require("x").default` when consuming the library from commonjs, but at least it's consistent. Best advise I can give: try to avoid the default export.

---

While we are here, there is another inconsistency when generating a commonjs bundle with rollup:

A module like `import x from "cjs"; console.log(x);` would result in the following interops generated:

``` js
// typescript
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cjs_1 = __importDefault(require("cjs"));
console.log(cjs_1.default);
```

``` js
// babel
"use strict";

var _cjs = _interopRequireDefault(require("cjs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

console.log(_cjs.default);
```

``` js
// webpack
/******/ 		// getDefaultExport function for compatibility with non-harmony modules
/******/ 		__webpack_require__.n = (module) => {
/******/ 			var getter = module && module.__esModule ?
/******/ 				() => module['default'] :
/******/ 				() => module;
/******/ 			Object.defineProperty(exports, "a", { enumerable: true, get: getter });
/******/ 			return getter;
/******/ 		};

/* harmony import */ var cjs__WEBPACK_IMPORTED_MODULE_0__ = require("x");
/* harmony import */ var cjs__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(cjs__WEBPACK_IMPORTED_MODULE_0__);

console.log((cjs__WEBPACK_IMPORTED_MODULE_0___default()));
```

``` js
// rollup
function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

console.log(_interopDefault(require('x')));
```

It's not obvious to see, but babel, typescript and webpack check the `__esModule` flag before accessing the `default` property of a commonjs module, while rollup only checks the existance of a default property. This would behave differently with a module like `module.exports = { default: 42 };`.

**Comment by lukastaegert** [comment]
That Rollup looks for default instead of __esModule is an unfortunate design decision from before my time because a user complained that interop was not working when an internally generated namespace object was attached to a global variable. The solution of course should rather have been to attach the __esModule prop to internal namespaces instead of changing interop, but alas, yes that should be fixed, though likely in a new major.

That aside, interop is indeed the crucial question. And the fact that the commonjs plugin also just checks for the presence of a default export should also be changed. My suggested interop above is indeed slightly different: require should only return the default if there are no other exports.

There are of course quite a few plus points for always returning the namespace on require:
- It allows for live binding of the default export
- It is apparently consistent with other tools
- Should Node ever support this, they are likely to also return the namespace

There is however one downside:
- A formerly pure CJS package that just exported e.g. a function via attaching to module.exports cannot add an ESM build with compatible interop. It’s just impossible.

And I am not sure this is good enough. I was hoping to get an answer from Webpack for this question which is better than: Please break your library interface in favour of a more complicated one.

**Comment by sokra** [comment]
Regarding the dual state hazard. There is actually a way to write your library in esm, transpile it to cjs, avoid visible double instantiation and the `import` condition would not have any commonjs module in the graph. It's inspired by the idea of isolating state in json modules proposed by someone in the other threads (was unable to find the comment, otherwise I would give credit here).

``` json
"exports": {
  "import": "./wrapper.mjs",
  "require": "./wrapper.cjs"
}
```

``` js
// wrapper.json
{}
```

``` js
// wrapper.mjs
import wrap from "./wrapper.json";
import * as lib from "./index.mjs";

if(!wrap.__esModule) {
  Object.assign(wrap, lib);
  Object.defineProperty(wrap, "__esModule", { value: true });
}

// Repeat the exports here:
export const A = wrap.A;
export const B = wrap.B;
export const C = wrap.C;
```

``` js
const wrap = require("./wrapper.json");

if(!wrap.__esModule) {
  Object.assign(wrap, require("./index.cjs"));
  Object.defineProperty(wrap, "__esModule", { value: true });
}

module.exports = wrap;
```

The wrapper would make sure that only the first requested version is exposed to the user.

`index.mjs` can include pure ESM, without any reference to CJS.

There is a very small runtime overhead when requesting the lib via `require()` before `import`. This is case library is executed twice, but still not exposed twice. Executing the library should not have a side effect otherwise that would break.

`index.cjs` can contain an exact transpiled version of `index.mjs`, without any extra modification.

Note that still would not solve the double bundle problem, but at least the double state problem at runtime. Bundler could detect this pattern to optimize instead of using the `moduleOnly` condition, but this sounds a bit complex. It also depends on user following the pattern closely and don't get to creative...

**Comment by sokra** [comment]
> There are of course quite a few plus points for always returning the namespace on require:
> 
> * It allows for live binding of the default export
> * It is apparently consistent with other tools
> * Should Node ever support this, they are likely to also return the namespace
> 
> There is however one downside:
> 
> * A formerly pure CJS package that just exported e.g. a function via attaching to module.exports cannot add an ESM build with compatible interop. It’s just impossible.

Yes that's an downside, and some people already wondered about that. In such a package adding ESM would need to be a breaking change. That would be something I personally would accept (or have accepted in the past) in favor of greater consistency in the whole system.

Allow single export default would only move the breaking change. Once you want to add another named export next to the default export this would mean a breaking change for the package.

**Comment by sokra** [comment]
Seems like we won't find a consensus here, but as each tool can decide this on its own, I go on and decide something for webpack here that I think is appropriate for webpack.

We will implement it this way in webpack 5 beta for now, but it's not set in stone. It's still a beta and we could still change this again.

Thanks for all the input you all have given. I think the webpack guideline contains a piece from everyone. You probably don't like something about that, I'm sorry for that, I also wished we found a 
consensus for that.

See a draft for the guideline here: https://gist.github.com/sokra/e032a0f17c1721c71cfced6f14516c62

Some notes on that:

* `module` instead of `moduleOnly`. It's not that strict, but I wrote the guideline in a way that tools that only support ESM can statically detect if `module` is valid according to this restrictions. I think that's better than having two conditions `module` and `moduleOnly` or only `moduleOnly`.
* It varies a little bit from the node guidelines and even defines some of them as anti-patterns. Sorry about that. The reasons for that are explained in the document.
